### PR TITLE
Fix conditional GFlowNet training: implement functional continuous conditioning

### PR DESCRIPTION
**Pull request by saleml**
- Introduced new parameters in ConditionalArgs for learning rate, validation interval, and exploration.
- Updated test scripts to include various conditional training scenarios and validation checks.
- Enhanced train_conditional.py with a new ConditionalHyperGrid class for conditioning-aware rewards.
- Implemented validation during training to assess performance across different conditioning values.
- Added evaluation function for detailed metrics on conditional sampling distributions.

- [x] I've read the [.github/CONTRIBUTING.md](.github/CONTRIBUTING.md) file
- [x] My code follows the typing guidelines
- [x] I've added appropriate tests
- [x] I've run pre-commit hooks locally

## Description
Addresses https://github.com/GFNOrg/torchgfn/issues/190
 ## Summary
  This PR fixes the non-functional conditioning in `train_conditional.py` and implements proper continuous conditioning
  that interpolates between uniform and multi-modal distributions.

  ## Problem
  The original conditioning implementation was decorative - conditioning values didn't actually affect the reward function,
   making the conditional GFlowNet behave identically regardless of conditioning.

  ## Changes

  ### Core Functionality
  - **Functional conditioning**: Conditioning now actually modifies rewards via linear interpolation
    - `conditioning=0`: Uniform distribution (all states get R0+R1+R2)
    - `conditioning=1`: Original HyperGrid multi-modal distribution
    - `conditioning∈(0,1)`: Linear interpolation between the two extremes

  ### Architecture Improvements
  - **LogZ depends only on conditioning**: Fixed LogZ to be a function of conditioning only (not states), matching the
  theoretical requirement for conditional GFlowNets
  - **Simplified validation**: Uses fresh trajectory samples for each conditioning value during validation
  - **Robust reward calculation**: Handles batch size mismatches that can occur during DetailedBalance loss calculation

  ### Training & Validation
  - Added hyperparameter arguments matching `train_hypergrid_simple.py`: `--lr`, `--lr_logz`, `--epsilon`,
  `--validation_samples`
  - Validation now samples fresh trajectories per conditioning value (doesn't reuse mixed-conditioning trajectories)
  - Concise logging: `Iter X: L1=[...] modes=Y` format
  - Mode discovery tracking for conditioning=1

  ### Testing
  - Updated test suite with functional tests (not just smoke tests)
  - Added tests for convergence, different dimensions, exploration, and all GFlowNet types
  - Fixed test compatibility issues with the `evaluate` flag

  ## Technical Details

  The key insight is that in conditional GFlowNets:
  - The reward function should vary based on conditioning
  - LogZ (partition function) should depend only on conditioning, not on states
  - During training, we sample conditioning uniformly from [0,1] for each batch

  ## Example Usage
  ```bash
  python train_conditional.py --ndim 2 --height 8 --epsilon 0.1 --lr 1e-3 --lr_logz 1e-2
```

  Testing

  All tests pass with the new implementation:
  - test_conditional_basic: Tests all GFlowNet types (TB, DB, SubTB, FM)
  - test_conditional_convergence: Verifies training converges to reasonable losses
  - test_conditional_different_dims: Tests with different environment dimensions
  - test_conditional_with_exploration: Tests with epsilon-greedy exploration
  - test_conditional_loss_types: Comprehensive test of all loss types

### CODE REVIEW

**Comment by saleml** [comment]
I don't see scenarios where logZ would depend on the state. In a non-conditional TB GFlowNets, logZ is simply a scalar, independent of the state. In a conditional one, it should be a function of the condition only. So perhaps we could make 'ConditionalLogZEstimator' as such and use it in this script ?

**Comment by josephdviviano** [comment]
That said, the n_modes metric is 0 at the end of training. I'm not sure if that's just because the "mode found" logic is overly strict (it can be because there is aliasing of the grid, i.e., the true mode is between two grid locations and the reward value is never superthreshold). Are you seeing this on your side?
